mrmustard/lab_dev/wires.py

```diff
@@ -200,22 +208,50 @@ def __add__(self, other: Wires) -> Wires:
         for m in all_modes:
             self_row = self.id_array[self._modes.index(m)] if m in self.modes else np.zeros(4)
             other_row = other.id_array[other._modes.index(m)] if m in other.modes else np.zeros(4)
-            # if np.any(np.where(self_row > 0) == np.where(other_row > 0)):
-            #     raise ValueError(f"wires overlap on mode {m}")
+            if np.any(np.where(self_row > 0) == np.where(other_row > 0)):
+                raise ValueError(f"wires overlap on mode {m}")
             modes_rows[m] = [s if s > 0 else o for s, o in zip(self_row, other_row)]
         combined_array = np.array([modes_rows[m] for m in sorted(modes_rows)])
-        return self._from_data(None, sorted(modes_rows), np.ones_like(combined_array))
         return self._from_data(combined_array, sorted(modes_rows), np.ones_like(combined_array))
 
     def __bool__(self) -> bool:
-        return True if self.ids else False
+        return True if len(self.ids) > 0 else False
 
     def __getitem__(self, modes: Iterable[int] | int) -> Wires:
         "A view of this Wires object with wires only on the given modes."
         modes = [modes] if isinstance(modes, int) else modes
         idxs = tuple(list(self._modes).index(m) for m in set(self._modes).difference(modes))
         return self._view(masked_rows=idxs)
 
+    def _mode(self, mode: int) -> np.ndarray:
+        "A slice of the id_array matrix at the given mode."
+        return np.maximum(0, self.id_array[[self._modes.index(mode)]])[0]
+
+    def __lshift__(self, other: Wires) -> Wires:
+        return (other.dual >> self.dual).dual  # how cool is this
+
+    def __rshift__(self, other: Wires) -> Wires:
+        r"""Returns a new Wires object with the wires of self and other combined as two
+        components in a circuit: the output of self connects to the input of other wherever
+        they match.
+        """
+        all_modes = set(self.modes) | set(other.modes)
+        new_id_array = np.zeros((len(all_modes), 4), dtype=np.int64)
+        for i, m in enumerate(sorted(all_modes)):
+            if m in self.modes and m in other.modes:
+                if bool(self[m].output.ket) != bool(other[m].input.ket) or bool(
+                    self[m].output.bra
+                ) != bool(other[m].input.bra):
+                    raise ValueError(f"wire mismatch on mode {m}")
+                new_id_array[i] += np.array(
+                    [other._mode(m)[0], self._mode(m)[1], other._mode(m)[2], self._mode(m)[3]]
+                )
+            elif m in self.modes:
+                new_id_array[i] += self._mode(m)
+            elif m in other.modes:
+                new_id_array[i] += other._mode(m)
+        return self._from_data(np.abs(new_id_array), sorted(all_modes), np.ones_like(new_id_array))
```

Comment:
```suggestion
        return self._from_data(np.abs(new_id_array), all_modes)
```
- `sorted(all_modes) --> all_modes`: as above
- np.ones_like: Not needed, it's default